### PR TITLE
balancemagic: rollup of changes requested by balance team

### PR DESCRIPTION
**Pull request by Kazinsal**
Includes splitting of fighter balance as per Durandal's request, and some mostly working but dummied-out-via-configuration burstfire code. I need to spend some time working on that one a bit more (the fighter balance split works fine though).

### CODE REVIEW

File: Plugins/Public/balancemagic/Main.cpp [hunk 1]
```diff
@@ -81,10 +90,11 @@ sockaddr_in sockaddrFLHook;
 sockaddr_in sockaddrKillposter;
 SOCKET sKillposterSocket = NULL;
 
-map<uint, CLIENT_DATA> mapKillmailData;
+CLIENT_DATA aClientData[250];
```
Comment (dsyalex) [inline Plugins/Public/balancemagic/Main.cpp]: isn't there a constant for this?

File: Plugins/Public/balancemagic/Main.cpp [hunk 2]
```diff
@@ -164,15 +177,16 @@ void LoadSettings()
 				{
 					struct DamageAdjust stEntry = { 0.0f };
 					stEntry.fighter = ini.get_value_float(0);
-					stEntry.transport = ini.get_value_float(1) ? ini.get_value_float(1) : stEntry.fighter;
-					stEntry.gunboat = ini.get_value_float(2) ? ini.get_value_float(2) : stEntry.transport;
-					stEntry.cruiser = ini.get_value_float(3) ? ini.get_value_float(3) : stEntry.gunboat;
-					stEntry.battlecruiser = ini.get_value_float(4) ? ini.get_value_float(4) : stEntry.cruiser;
-					stEntry.battleship = ini.get_value_float(5) ? ini.get_value_float(5) : stEntry.battlecruiser;
+					stEntry.freighter = ini.get_value_float(1) ? ini.get_value_float(1) : stEntry.fighter;
```
Comment (dsyalex) [inline Plugins/Public/balancemagic/Main.cpp]: so this will be backwards-incompatible with old inis?

File: Plugins/Public/balancemagic/Main.cpp [hunk 3]
```diff
@@ -455,23 +487,151 @@ void __stdcall HkCb_AddDmgEntry(DamageList *dmg, unsigned short p1, float damage
...
+
+	if (!iLoadedBurstFires)
+		return;
+
+	IObjInspectImpl *obj = HkGetInspect(iClientID);
+	if (obj)
+	{
+		CShip* cship = (CShip*)HkGetEqObjFromObjRW((IObjRW*)obj);
+
+		for (short* slot = wpn.sArray1; slot != wpn.sArray2; slot++)
```
Comment (dsyalex) [inline Plugins/Public/balancemagic/Main.cpp]: can we get sArray1/sArray2 named?

File: Plugins/Public/balancemagic/Main.cpp [hunk 4]
```diff
@@ -455,23 +487,151 @@ void __stdcall HkCb_AddDmgEntry(DamageList *dmg, unsigned short p1, float damage
...
+			if (CEGun::cast(equip))
+			{
+				CEGun* cgun = (CEGun*)equip;
+				map<unsigned int, struct AmmoGenerator>::iterator iter = mapBurstFire.find(cgun->GunArch()->iArchID);
+				if (iter != mapBurstFire.end())
+				{
+					aClientData[iClientID].mapBurstFills[*slot].ammo = iter->second.ammo;
+					aClientData[iClientID].mapBurstFills[*slot].count++;
+					aClientData[iClientID].mapBurstFills[*slot].timestamp = GetTimeInMS() + iter->second.timestamp;
+					ConPrint(L"BURSTFIRE: Client %u fired a round of 0x%08X from slot %u.\n", iClientID, iter->second.ammo, *slot);
```
Comment (dsyalex) [inline Plugins/Public/balancemagic/Main.cpp]: is this something we want to print potentially very often?

File: Plugins/Public/balancemagic/Main.cpp [hunk 5]
```diff
@@ -455,23 +487,151 @@ void __stdcall HkCb_AddDmgEntry(DamageList *dmg, unsigned short p1, float damage
...
+		CEquip *equip = eqmanager->Traverse(tr);
+		while (equip)
+		{
+			if (CEGun::cast(equip))
+			{
+				CEGun* cgun = (CEGun*)equip;
+				map<unsigned int, struct AmmoGenerator>::iterator iter = mapBurstFire.find(cgun->GunArch()->iArchID);
+				if (iter != mapBurstFire.end())
+				{
+					ConPrint(L"BURSTFIRE: PlayerLaunch_AFTER - Iteration %d for client %u, removing 200 units of 0x%08X.\n", i, iClientID, iter->second.ammo);
```
Comment (dsyalex) [inline Plugins/Public/balancemagic/Main.cpp]: ^

File: Plugins/Public/balancemagic/Main.cpp [hunk 6]
```diff
@@ -455,23 +487,151 @@ void __stdcall HkCb_AddDmgEntry(DamageList *dmg, unsigned short p1, float damage
...
+				{
+					ammogenerator[iter->second.ammo] += iter->second.count;
+					aClientData[i].mapBurstFills.erase(iter->first);
+				}
+			}
+
+			for (map<uint, uint>::iterator iter = ammogenerator.begin(); iter != ammogenerator.end(); ++iter)
+			{
+				HkAddCargo((const wchar_t*)Players.GetActiveCharacterName(i), iter->first, iter->second, false);
+				ConPrint(L"BURSTFIRE: Tick - Gave client %u %u rounds of 0x%08X.\n", i, iter->second, iter->first);
```
Comment (dsyalex) [inline Plugins/Public/balancemagic/Main.cpp]: print on tick?

File: Plugins/Public/balancemagic/Main.cpp [hunk 7]
```diff
@@ -455,23 +487,151 @@ void __stdcall HkCb_AddDmgEntry(DamageList *dmg, unsigned short p1, float damage
...
+void HkTimerNPCAndF1Check()
+{
+	returncode = DEFAULT_RETURNCODE;
+
+	if (!iLoadedBurstFires)
+		return;
+
+	mstime curr_mstime = GetTimeInMS();
+	//ConPrint(L"BURSTFIRE: Tick %llu.\n", curr_mstime);
+	for (int i = 0; i < 250; i++)
```
Comment (dsyalex) [inline Plugins/Public/balancemagic/Main.cpp]: same as earlier with the max client ID thing

File: Plugins/Public/balancemagic/Main.cpp (hunk 2 above)
Comment (Kazinsal) [reply Plugins/Public/balancemagic/Main.cpp]: Correct. This was done at the request of the balance team. I have submitted to them an updated config based on the live one,

File: Plugins/Public/balancemagic/Main.cpp (hunk 1 above)
Comment (Kazinsal) [reply Plugins/Public/balancemagic/Main.cpp]: Every plugin handles client data differently.

File: Plugins/Public/balancemagic/Main.cpp (hunk 3 above)
Comment (Kazinsal) [reply Plugins/Public/balancemagic/Main.cpp]: Probably could but that's an FLHook core thing, not a this-plugin thing

File: Plugins/Public/balancemagic/Main.cpp (hunk 4 above)
Comment (Kazinsal) [reply Plugins/Public/balancemagic/Main.cpp]: Yeah this is in my testing bits but the whole thing is dummied out at the moment so it doesn't matter

File: Plugins/Public/balancemagic/Main.cpp (hunk 5 above)
Comment (Kazinsal) [reply Plugins/Public/balancemagic/Main.cpp]: Likewise

File: Plugins/Public/balancemagic/Main.cpp (hunk 7 above)
Comment (Kazinsal) [reply Plugins/Public/balancemagic/Main.cpp]: It's somewhere in there, I'll go dig it up

File: Plugins/Public/balancemagic/Main.cpp (hunk 6 above)
Comment (Kazinsal) [reply Plugins/Public/balancemagic/Main.cpp]: resolved in commit 2

**Comment by Kazinsal** [comment]
> have not yet checked logic or HkCb_AddDmgEntry priority change

priority change actually fixes an edge case where base.dll would override balancemagic.dll. logic in active code is the same as the previous code with just a few lines changed.

the balance team wanted this done. I have done it. there just happens to be some other stuff I was working on for them in there that I've deactivated because it's not prod-ready yet.

**Pull request merged**